### PR TITLE
bugfix: py contains raises errors when appropiate

### PR DESCRIPTION
**Pull request by Skylion007**
## Description
* Previously our contains functions just swallowed errors like TypeErrors for unhashable types. This is poorly defined and only really worked with dict and sets (but set the Python API in a bad state).  @rwgk Could you do some global testing to see how widespread the problem is off using contains to check if a key is hashable.
* There is a similar problem with PySet_Add(), but we abuse that ourselves in the STL casters, so we will need some utilities to check if an object is hashable first.

## Suggested changelog entry:

```rst
* Properly raise exceptions in contains methods (like when an object in unhashable).
```

### CODE REVIEW

**Comment by rwgk** [comment]
> Could you do some global testing to see how widespread the problem is off using contains to check if a key is hashable.

Will do.

**Comment by rwgk** [comment]
Hi @mattip, PyPy doesn't seem to raise `TypeError` as expected in the added tests, while all cpython versions do. Is that an intentional difference in behavior?

```
def test_unhashable_exceptions(arg, func):
    class Unhashable:
        __hash__ = None

    with pytest.raises(TypeError) as exc_info:
        func(arg, Unhashable())
```

**Comment by mattip** [comment]
If I understand correctly, `func` here is `PySet_Contains`, `PyDict_Contains`, `class.__contains__()` ? I can reproduce the PyPy problem for `PyDict_Contains` but not for the other two. I fixed `PyDict_Contains` in [PyPy 99677b3cad6e](https://foss.heptapod.net/pypy/pypy/-/commit/9e29512d11397f190d1ed4ac9064d33f68de2e27).

**Comment by rwgk** [comment]
> If I understand correctly, `func` here is `PySet_Contains`, `PyDict_Contains`, `class.__contains__()` ? I can reproduce the PyPy problem for `PyDict_Contains` but not for the other two. I fixed `PyDict_Contains` in [PyPy 99677b3cad6e](https://foss.heptapod.net/pypy/pypy/-/commit/9e29512d11397f190d1ed4ac9064d33f68de2e27).

Thanks @mattip!

I downloaded and unpacked the "log archive" [here](https://github.com/pybind/pybind11/actions/runs/3176232918/jobs/5175291993), then:

```
$ grep 'arg =' *.txt
12______pypy-3.7_____windows-2022_____x64.txt:2022-10-03T17:50:59.4914903Z   arg = set(), func = <builtin_function_or_method object at 0x000001e499692950>
13______pypy-3.8_____windows-2022_____x64.txt:2022-10-03T17:50:45.7048591Z   arg = set(), func = <builtin_function_or_method object at 0x0000025a0a13fec0>
14______pypy-3.9_____windows-2022_____x64.txt:2022-10-03T17:51:14.2422408Z   arg = set(), func = <builtin_function_or_method object at 0x0000020df0abc100>
19______pypy-3.7_____macos-latest_____x64.txt:2022-10-03T18:01:25.6466500Z arg = set(), func = <builtin_function_or_method object at 0x00007f866f4be480>
20______pypy-3.8_____macos-latest_____x64.txt:2022-10-03T17:59:04.6502290Z arg = set(), func = <builtin_function_or_method object at 0x00007fa078f29600>
21______pypy-3.9_____macos-latest_____x64.txt:2022-10-03T17:55:13.3994880Z arg = set(), func = <builtin_function_or_method object at 0x00007fec66e64800>
5______pypy-3.7_____ubuntu-latest_____x64.txt:2022-10-03T17:49:03.7385057Z arg = set(), func = <builtin_function_or_method object at 0x00000000021403a0>
6______pypy-3.8_____ubuntu-latest_____x64_-DPYBIND11_FINDPYTHON=ON.txt:2022-10-03T17:48:25.9971616Z arg = set(), func = <builtin_function_or_method object at 0x00000000027e54b0>
7______pypy-3.9_____ubuntu-latest_____x64.txt:2022-10-03T17:49:23.3291437Z arg = set(), func = <builtin_function_or_method object at 0x00007fa78b01bd00>
```

AFAICT it only fails for `set()`, but succeeds for `dict()` and `CustomContains()` (I'm sure pytest runs all 3 even if one of them fails), for all PyPy versions that we cover.

@Skylion007 does my understanding look right to you?

**Comment by mattip** [comment]
Right, I [can reproduce](https://github.com/pypy/binary-testing/actions/runs/3182020672/jobs/5187448698#step:12:92) with a nightly pypy in a repo that runs CI on popular C extensions. I will try to fix it there.

**Comment by rwgk** [comment]
> > Could you do some global testing to see how widespread the problem is off using contains to check if a key is hashable.
> 
> Will do.

I don't see any problems in the global testing results related to this PR.

Very minor caveat: results are unusually noisy this morning, but I looked pretty carefully.

**Comment by mattip** [comment]
There is a corner case for empty sets where PyPy does not raise. CPython2 and CPython3 raise. See https://foss.heptapod.net/pypy/pypy/-/issues/3824

File: tests/test_pytypes.py [hunk 1]
```diff
@@ -175,16 +175,12 @@ def __contains__(self, m):
         return m in self.d
 
 
+@pytest.mark.xfail("env.PYPY and sys.pypy_version_info < (7, 3, 7)")
```
Comment (mattip) [inline tests/test_pytypes.py:178]:
```suggestion
@pytest.mark.xfail("env.PYPY and sys.pypy_version_info < (7, 3, 10)")
```

Comment (mattip) [reply tests/test_pytypes.py:178]: Currently PyPy has releases 7.3.9, the next will be 7.3.10

File: include/pybind11/pytypes.h [hunk 2]
```diff
@@ -2349,7 +2357,11 @@ args_proxy object_api<D>::operator*() const {
 template <typename D>
 template <typename T>
 bool object_api<D>::contains(T &&item) const {
-    return attr("__contains__")(std::forward<T>(item)).template cast<bool>();
+    auto ret = attr("__contains__")(std::forward<T>(item)).template cast<bool>();
+    if (PyErr_Occurred()) {
```
Comment (rwgk) [inline include/pybind11/pytypes.h:2361]: Can this actually happen?

I was thinking the line before will either succeed or there will be an exception (`error_already_set`, `cast_error`) in flight already.

Comment (Skylion007) [reply include/pybind11/pytypes.h]: @rwgk Good catch, my test designed to test this pathway fired fine without the if statement locally.

**Review by rwgk — APPROVED** [review]
Thanks Aaron for plugging these holes, and the new tests!

**Comment by rwgk** [comment]
CentOS7 / PGI 22.3  appears to be generally broken (same failure in another PR) and the other failing test as a notorious flake. I think it's best if we ignore the CentoOS failure, merge our PRs, then one of us can look what's going on with CentOS.

**Comment by rwgk** [comment]
@Skylion007 I just rebased this on master, to trigger the CI, to see if the CentOS7 / PGI 22.3 problem was fixed in the meantime.

**Comment by rwgk** [comment]
@henryiii @ax3l Do you know what's behind the error below? First observed yesterday, still failing now.

```
Downloading packages:
Package nvhpc-22.3-1.x86_64.rpm is not signed
--------------------------------------------------------------------------------
Total                                               59 MB/s | 5.9 GB  01:42     
warning: /var/cache/yum/x86_64/7/nvhpc/packages/nvhpc-22-9-22.9-1.x86_64.rpm: Header V4 RSA/SHA512 Signature, key ID 264a7796: NOKEY
Retrieving key from https://developer.download.nvidia.com/hpc-sdk/rhel/RPM-GPG-KEY-NVIDIA-HPC-SDK
Importing GPG key 0x264A7796:
 Userid     : "NVIDIA HPC SDK"
 Fingerprint: 5793 edb8 2f7b fb26 ca4b 783f 7343 6a89 264a 7796
 From       : https://developer.download.nvidia.com/hpc-sdk/rhel/RPM-GPG-KEY-NVIDIA-HPC-SDK


Package nvhpc-22-3-22.3-1.x86_64.rpm is not signed
Error: Process completed with exit code 1.
```

**Comment by rwgk** [comment]
Hi @Skylion007, is there a chance that you could merge this today? (Asking b/o I'm hoping to include this in the next smart_holder update.)

**Comment by Skylion007** [comment]
@rwgk Wasn't sure if wanted to wait until the next major update as this is a bit of a behavior change.

**Comment by Skylion007** [comment]
If we are good with it, feel free to merge it.

**Review by henryiii — APPROVED** [review]
Yes, I’d say this is a bug fix.

**Comment by rwgk** [comment]
Thanks!

**Comment by Skylion007** [comment]
Only concern is that set.add() is still broken, and we sort of rely on that broken behavior in the caster.

**Comment by rwgk** [comment]
Sorry, looks like I was a bit fast.

> So that's a bigger PR issue I haven't looked at.

Leaving that for later sounds like the right move, does that make sense?
